train.py

```diff
@@ -336,10 +336,14 @@ def loss_fn(pred, labels):
             ):
                 losses = [loss.item() for loss in losses_since_last_log]
                 avg_loss, max_loss = sum(losses) / len(losses), max(losses)
-                if parallel_dims.dp_enabled:
+                if (
+                    parallel_dims.dp_replicate_enabled
+                    or parallel_dims.dp_shard_enabled
+                    or parallel_dims.cp_enabled
+                ):
                     global_avg_loss, global_max_loss = (
-                        utils.dist_mean(avg_loss, dp_mesh),
-                        utils.dist_max(max_loss, dp_mesh),
+                        utils.dist_mean(avg_loss, world_mesh["dp_cp"]),
```

Comment:
something to consider for the future:
constructing NCCL communicators may use extra GPU memory
depending on the tradeoff, doing 2 separate all-reduces over existing communicators may some memory over creating a new communicator to do a single all-reduce

Reply:
Makes sense. Tbh I'd hope device mesh supports all-reduce over a 2-D mesh and handles that for me, but maybe this use case is not be popular enough.